src/dataflow-types/src/client.rs

```diff
@@ -964,3 +1122,248 @@ pub mod process_local {
+        }
+    }
+
+    #[async_trait]
+    impl<T> GenericClient<Command<T>, Response<T>> for TcpClient<T>
+    where
+        T: timely::progress::Timestamp + Copy + Unpin,
+    {
+        async fn send(&mut self, cmd: Command<T>) -> Result<(), anyhow::Error> {
+            // TODO: something better than panicking.
+            self.connection.send(cmd).await.map_err(|err| err.into())
+        }
+
+        async fn recv(&mut self) -> Option<Response<T>> {
+            // TODO: something better than panicking.
+            self.connection
+                .next()
+                .await
+                .map(|x| x.expect("connection to dataflow server broken"))
+        }
+    }
+
+    /// A framed connection to a dataflowd server.
+    pub type Framed<C, T, U> = tokio_serde::Framed<
+        tokio_util::codec::Framed<C, LengthDelimitedCodec>,
+        T,
+        U,
+        Bincode<T, U>,
+    >;
+
+    /// A framed connection from the server's perspective.
+    pub type FramedServer<C, T> = Framed<C, Command<T>, Response<T>>;
+
+    /// A framed connection from the client's perspective.
+    pub type FramedClient<C, T> = Framed<C, Response<T>, Command<T>>;
+
+    fn length_delimited_codec() -> LengthDelimitedCodec {
+        // NOTE(benesch): using an unlimited maximum frame length is problematic
+        // because Tokio never shrinks its buffer. Sending or receiving one large
+        // message of size N means the client will hold on to a buffer of size
+        // N forever. We should investigate alternative transport protocols that
+        // do not have this limitation.
+        let mut codec = LengthDelimitedCodec::new();
+        codec.set_max_frame_length(usize::MAX);
+        codec
+    }
+
+    /// Constructs a framed connection for the server.
+    pub fn framed_server<C, T>(conn: C) -> FramedServer<C, T>
+    where
+        C: AsyncRead + AsyncWrite,
+    {
+        tokio_serde::Framed::new(
+            tokio_util::codec::Framed::new(conn, length_delimited_codec()),
+            Bincode::default(),
+        )
+    }
+
+    /// Constructs a framed connection for the client.
+    pub fn framed_client<C, T>(conn: C) -> FramedClient<C, T>
+    where
+        C: AsyncRead + AsyncWrite,
+    {
+        tokio_serde::Framed::new(
+            tokio_util::codec::Framed::new(conn, length_delimited_codec()),
+            Bincode::default(),
+        )
+    }
+}
+
+/// A client composed of channels.
+mod channel {
+    use std::fmt;
+
+    use async_trait::async_trait;
+    use derivative::Derivative;
+
+    use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+
+    use crate::client::GenericClient;
+
+    /// A client composed of channels.
+    #[derive(Derivative)]
+    #[derivative(Debug)]
+    pub struct ChannelClient<C1, C2, R> {
+        tx: UnboundedSender<C2>,
+        rx: UnboundedReceiver<R>,
+        #[derivative(Debug = "ignore")]
+        command_map: Box<dyn Fn(C1) -> C2 + Send>,
```

Comment:
Nit, but could this be `FnMut`? I'm thinking of all the `Client` implementations I've written and they are largely stateful. Maybe that is not what this is for though.